Fix #1388: Allow Agora to restart

Geod24 opened the pull request:
First commit should be fairly obvious. Second and third one attempt to address this issue:
```
2020-11-28 21:22:48,94 Info [SCP] - LocalNode::LocalNode@GDNOD qSet: 204980
2020-11-28 21:22:48,94 Info [agora.consensus.SCPEnvelopeStore] - Loading database from: .cache/scp_envelopes.dat
core.exception.AssertError@source/agora/consensus/EnrollmentManager.d(649): UTXO for validator not found!
```

Geod24 committed:
Ledger: Use peekUTXO when updating validator set

The block has already been validated, so the UTXO has to be there.
But depending on the state of the case, it might be masked,
hence just use peekUTXO for simplicity.

Geod24 committed:
Only re-enroll validators if there is none

The Ledger constructor was always re-enrolling the last few validators,
which means restarting a node would trigger an assertion failure.
This critical bug is worked around by only triggering this piece of code
when there is no validator (which should never happen).
Later on, this logic should be moved to catch-up.

Geod24 committed:
Use peekUTXO when calling EM.isEnrolled

EnrollmentManager.isEnrolled requires an UTXOFinder, which in itself is okay,
but providing one with replay protection can be dangerous as it has side effects.

Geod24 commented:
This needs a test

AndrejMitrovic reviewed (approved):
LGTM. Feel free to add test-case or merge.

Geod24 committed:
Tests: Add a convenience template mixin to forward to node ctor

This is a common pattern, and we shouldn't have to copy paste code all over the place.

Geod24 committed:
Test: Pass TestConf to the node directly

This allows us to have the same signature for both constructors,
which we will take advantage of in a later improvement.

Geod24 committed:
Test: Add convenience 'addNewNode' function to APIManager

Once again to simplify a common pattern, this will reduce the amount of boilerplate
that has to be put into createNewNode.

Geod24 committed:
Add a restart test that use persistence

Make sure that when we restart and there is data on disk,
we don't outright crash, as that's not the expected behavior.

Geod24 commented:
I have a test case locally but this needs more work:
```
2020-11-30 18:47:21,858 Info [agora.network.NetworkManager] - Retrieving blocks [1..1] from Validator #1 (GDNODE3EWQKF33TPK35DAQ3KXAYSOT4E4ACDOVJMDZQDVKP66IMJEACM)..
2020-11-30 18:47:21,858 Info [agora.network.NetworkManager] - Received blocks [1..1]
2020-11-30 18:47:21,858 Trace [agora.consensus.state.UTXOSet] - findUTXO: utxo_hash 0x6d85d61fd9d7bb663349ca028bd023ad1bd8fa65c68b4b1363a9c7406b4d663fd73fd386195ba2389100b5cd5fc06b440f053fe513f739844e2d72df302e8ad0 not found
2020-11-30 18:47:21,858 Trace [agora.node.Ledger] - Rejected block: Transaction: Input ref not in UTXO: Height: 1, Prev: 0x72e6...3b7d, Root: 0x4542...b1c8, Enrollments: [],
Transactions: 8
Type : Payment, Inputs (1): 0x6d85...8ad0:0x02b2...feb7
Outputs (1): GCOQ...LRIJ(61,000,000)
Type : Payment, Inputs (1): 0xc378...d314:0x0fbf...ba74
Outputs (1): GCOQ...LRIJ(61,000,000)
Type : Payment, Inputs (1): 0x451a...f0eb:0x0cbf...9070
Outputs (1): GCOQ...LRIJ(61,000,000)
Type : Payment, Inputs (1): 0xd446...cb92:0x0db4...6a07
Outputs (1): GCOQ...LRIJ(61,000,000)
Type : Payment, Inputs (1): 0x7e19...d685:0x0847...80d9
Outputs (1): GCOQ...LRIJ(61,000,000)
Type : Payment, Inputs (1): 0x3451...1e85:0x0348...0ecf
Outputs (1): GCOQ...LRIJ(61,000,000)
Type : Payment, Inputs (1): 0xfca9...baf8:0x02da...95d1
Outputs (1): GCOQ...LRIJ(61,000,000)
Type : Payment, Inputs (1): 0xff05...beeb:0x052b...09b1
Outputs (1): GCOQ...LRIJ(61,000,000)
```

Geod24 committed:
Ledger: Add an info message with the last known block

Geod24 committed:
MemBlockStorage: Allow load to be called multiple time

It would otherwise throw.

Geod24 commented:
Fixed, should look much better now
